cli/CommandExecutor.cpp

```diff
@@ -0,0 +1,200 @@
+    executeDescribeDatabase(arguments, catalog_database, out);
+  } else if (command_str.compare("\\d") == 0) {
+    if (arguments->size() == 0) {
+      executeDescribeDatabase(arguments, catalog_database, out);
+    } else {
+      executeDescribeTable(arguments, catalog_database, out);
+    }
+  } else {
+    THROW_SQL_ERROR_AT(command.command()) << "Invalid Command";
+  }
+}
+
+void CommandExecutor::executeDescribeDatabase(
+    const PtrVector<ParseString> *arguments,
+    const CatalogDatabase *catalog_database, FILE *out) {
+  std::size_t relation_count = 0;
+  // Column width initialized to 6 to take into account the header name
+  // and the column value table
+  int max_column_width = 6;
+  const CatalogRelation *relation;
+  if (arguments->size() == 0) {
+    for (auto rel_it = catalog_database->begin();
+         rel_it != catalog_database->end(); ++rel_it) {
+      int column_width = static_cast<int>(rel_it->getName().length());
+      max_column_width =
+          max_column_width < column_width ? column_width : max_column_width;
+      relation_count++;
+    }
+  } else {
+    const ParseString &table_name = arguments->front();
+    const std::string &table_name_val = table_name.value();
+    relation = catalog_database->getRelationByName(table_name_val);
+    int column_width = static_cast<int>(relation->getName().length());
+    max_column_width =
+        max_column_width < column_width ? column_width : max_column_width;
+    relation_count++;
+  }
+  // only if we have relations
+  if (relation_count > 0) {
+    vector<int> column_widths;
+    column_widths.reserve(relation_count);
+    column_widths.push_back(max_column_width);
+    column_widths.push_back(max_column_width);
+    fputs("       List of relations\n\n", out);
+    fprintf(out, "%-*s|", max_column_width, " Name");
+    fprintf(out, "%-*s|", 6, " Type");
+    fputc('\n', out);
+    PrintToScreen::printHBar(column_widths, out);
+    //  if there are no argument print the entire list of tables
+    //  else print the particular table only
+    if (arguments->size() == 0) {
+      for (auto rel_it = catalog_database->begin();
+           rel_it != catalog_database->end(); ++rel_it) {
+        fprintf(out, " %-*s|", max_column_width, rel_it->getName().c_str());
+        fprintf(out, " %-*s", max_column_width, "table");
+        fputc('\n', out);
+      }
+    } else {
+      fprintf(out, " %-*s|", max_column_width, relation->getName().c_str());
+      fprintf(out, " %-*s", max_column_width, "table");
+      fputc('\n', out);
+    }
+    fputc('\n', out);
+  }
+}
+
+void CommandExecutor::executeDescribeTable(
+    const PtrVector<ParseString> *arguments,
+    const CatalogDatabase *catalog_database, FILE *out) {
+  const ParseString &table_name = arguments->front();
+  const std::string &table_name_val = table_name.value();
+  const CatalogRelation *relation =
+      catalog_database->getRelationByName(table_name_val);
+  vector<int> column_widths;
```

Comment:
@zuyu static void printHBar(const std::vector<int> &column_widths,
                        FILE *out); 

takes a vector of integers.  This is an existing method in PrintToScreen. Should we change this method to take a vector of std::size_t instead of integers?